build/source/dshare/outpt_stat.f90

```diff
@@ -0,0 +1,294 @@
+    class default;err=20;message=trim(message)//'stat type not found';return
+   endselect
+
+   if (trim(meta(iVar)%varName).eq.'time') then
+    selecttype (stat)
+     typeis (dlength)
+      stat(iVar)%dat(iLookStat%inst) = tdata
+     class default;err=20;message=trim(message)//'time stat must be dlength';return
+    endselect
+   endif
+
+   if(err/=0)then; message=trim(message)//trim(cmessage);return; endif  
+  endif
+ enddo                                             ! model variables
+
+ return
+ end subroutine calcStats
+
+ ! ***********************************************************************************
+ ! Private subroutine calc_stats is a generic fucntion to deal with any variable type.
+ ! Called from compile_stats 
+ ! ***********************************************************************************
+ subroutine calc_stats(meta,stat,tdata,iStep,err,message)
+ USE nrtype
+ ! data structures
+ USE data_types,only:var_info,ilength,dlength ! type dec for meta data structures 
+ USE var_lookup,only:maxVarStat       ! # of output statistics 
+ USE globalData,only:outFreq          ! output frequencies 
+ ! structures of named variables
+ USE var_lookup,only:iLookVarType     ! named variables for variable types 
+ USE var_lookup,only:iLookStat        ! named variables for output statistics types 
+ implicit none
+ ! dummy variables
+ class(var_info),intent(in)        :: meta        ! meta dat a structure
+ class(*)       ,intent(inout)     :: stat        ! statistics structure
+ real(dp)       ,intent(in)        :: tdata       ! data structure
+ integer(i4b)   ,intent(in)        :: iStep       ! timestep
+ integer(i4b)   ,intent(out)       :: err         ! error code
+ character(*)   ,intent(out)       :: message     ! error message
+ ! internals
+ real(dp),dimension(maxvarStat+1)  :: tstat       ! temporary stats vector
+ integer(i4b)                      :: iStat       ! statistics loop
+ integer(i4b)                      :: iFreq       ! statistics loop
+ integer(i4b)   ,parameter         :: modelTime=1 ! model timestep
+ ! initialize error control
+ err=0; message='calc_stats/'
+
+ ! pull current frequency for normalization
+ iFreq = meta%outFreq
+
+ ! pack back into struc
+ selecttype (stat)
+  typeis (ilength); tstat = real(stat%dat)
+  typeis (dlength); tstat = stat%dat
+  class default;err=20;message=trim(message)//'stat type not found';return
+ endselect
+
+ ! ---------------------------------------------
+ ! reset statistics at new frequenncy period 
+ ! ---------------------------------------------
+ if ((mod(iStep,outFreq(iFreq)).eq.1).or.(outFreq(iFreq).eq.1)) then
+  do iStat = 1,maxVarStat                          ! loop through output statistics
+   if (.not.meta%statFlag(iStat)) cycle            ! don't bother if output flag is off
+   if (meta%varType.ne.iLookVarType%scalarv) cycle ! only calculate stats for scalars 
+   select case(iStat)                              ! act depending on the statistic 
```

Comment:
The reason that I did it this way was because it allowed me to make better line-by-line comments. Also, notice that the variance case requires two lines. Personally, I think it's cleaner this way, but if you disagree, I can change it. No big deal.